### PR TITLE
PXC-3760: Implement a feature to evict a lagging node from PXC

### PR DESCRIPTION
**Pull request by voodam**
https://jira.percona.com/browse/PXC-3760

Problem:
We want to evict a node if it slows down a whole cluster due to the flow control.

Solution:
We introduce two variables: gcs.fc_auto_evict_window and gcs.fc_auto_evict_threshold. If the node pauses in FC more then "threshold * window" in the last "window" seconds frame, node will just abort.

The abortion can be problematic if we have 2-nodes cluster. From the side of other nodes this will be indistinguishable from the crash of the lagging node.

### CODE REVIEW

**Comment by voodam** [comment]
I'll write the code to gcs_param_set soon, now the new params can't be dynamically changed.

Also, do we need to describe the new params somewhere (in readme or something)?

**Comment by voodam** [comment]
I found a problem with `pauses_window` allocation, fixing it.

**Comment by voodam** [comment]
The commit above is just a dirty fix, I probably need clarification here.

Do we strive to minimize allocations amount and hence `gcs_sm_t` shouldn't have any pointers (see `gcs_sm_create`)? I can use array and implement ring buffer, for example.

Or pointer is OK, but we need to use gu_malloc + placement-new to account allocation stats (see `galerautils/src/gu_mem.h`)?

File: gcs/src/gcs_sm.cpp [hunk 1]
```diff
@@ -137,9 +137,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
 {
     gu_mutex_destroy(&sm->lock);
     gu_cond_destroy(&sm->cond);
+    delete sm->pauses_window;
     gu_free (sm);
 }
 
+double
+gcs_sm_paused_in_window_get (gcs_sm_t* const sm, double window_sec,
```
Comment (kamil-holubicki) [inline gcs/src/gcs_sm.cpp]: `window_sec` is a rather constant value (changes only if the user changes `gcs.fc_auto_evict_window`), so maybe it would be better to avoid this multiplication every time?

File: gcs/src/gcs_sm.cpp [hunk 2]
```diff
@@ -137,9 +137,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+                             bool erase_outdated)
+{
+    assert(sm->fc_auto_evict);
+    long long now = gu_time_monotonic();
+    long long window_ns = now - window_sec * 1000000000LL;
```
Comment (kamil-holubicki) [inline gcs/src/gcs_sm.cpp]: window_ns -> window_start_ns

File: gcs/src/gcs_sm.cpp [hunk 3]
```diff
@@ -137,9 +137,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+        i = size - 2;
+    // not paused => num pauses == num conts, just start from the end
+    } else {
+        i = size - 1;
+    }
+    assert(i % 2); // there are pairs of pause and cont moments
+
+    for (; i > 0 && window[i] > window_ns; i -= 2) {
+        long long cont = window[i];
+        long long pause = std::max(window[i-1], window_ns);
```
Comment (kamil-holubicki) [inline gcs/src/gcs_sm.cpp]: I'm not sure if std::max is needed here. In the worst case, we will have an even larger calculated total_ns, which is fine, because later we just care about if the threshold was crossed.

File: gcs/src/gcs_sm.cpp [hunk 4]
```diff
@@ -137,9 +137,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+
+    /* Remove outdated moments. We need to move all fresh elements, but
+       it seems good anyway, because window probably won't be big. */
+    if (erase_outdated && i > 0) {
+        window.erase(window.begin(), window.begin()+i+1);
+        assert(window.size() % 2 == sm->pause);
+    }
+
+    gu_mutex_unlock (&sm->lock);
+    return total_ns * 1.0e-9;
```
Comment (kamil-holubicki) [inline gcs/src/gcs_sm.cpp]: Similarly to `window_sec`, it looks like it is possible to avoid frequent multiplications.

File: gcs/src/gcs.cpp [hunk 5]
```diff
@@ -2692,6 +2741,7 @@ _set_sync_donor (gcs_conn_t* conn, const char* value)
 
         conn->params.sync_donor = sd;
         conn->max_fc_state      = sd ? GCS_CONN_DONOR : GCS_CONN_JOINED;
+        gu_config_set_bool (conn->config, GCS_PARAMS_SYNC_DONOR, sd);
```
Comment (kamil-holubicki) [inline gcs/src/gcs.cpp]: why this change?

File: gcs/src/gcs.cpp [hunk 6]
```diff
@@ -2711,11 +2761,7 @@ _set_pkt_size (gcs_conn_t* conn, const char* value)
 
         long ret = gcs_set_pkt_size (conn, pkt_size);
 
-        if (ret >= 0)
-        {
-            ret = 0;
-            gu_config_set_int64(conn->config,GCS_PARAMS_MAX_PKT_SIZE,pkt_size);
-        }
+        if (ret >= 0) ret = 0;
```
Comment (kamil-holubicki) [inline gcs/src/gcs.cpp]: why this change?

**Review by kamil-holubicki — COMMENTED** [review]
Generally, it looks fine. Please see my comments.

Please do not refactor upstream code. Adding comments is fine, but formatting change, removing unused code, etc will cause problems in the future.

File: gcs/src/gcs_sm.hpp [hunk 7]
```diff
@@ -60,6 +58,12 @@ typedef struct gcs_sm
     long          cc;
 #endif /* GCS_SM_CONCURRENCY */
     bool          pause;
+    bool          fc_auto_evict; // is FC auto eviction enabled
+    /*
+      Sequence of pause and continue moments (first element is always pause),
+      eventually truncated to last GCS_PARAMS_FC_AUTO_EVICT_WND seconds.
+     */
+    std::vector<long long> * pauses_window;
```
Comment (kamil-holubicki) [inline gcs/src/gcs_sm.hpp]: What is wrong with having this vector as the member (not a pointer)?

Comment (voodam) [reply gcs/src/gcs_sm.hpp]: We have custom allocation in `gcs_sm_create`, so we should have size of the vector before calling this function. It means we should either add static size structures in `gcs_sm_t` only or have pointers to dynamic.

Please see also: https://github.com/percona/galera/pull/261#issuecomment-1488253731

File: gcs/src/gcs.cpp (hunk 6 above)
Comment (voodam) [reply gcs/src/gcs.cpp]: First time I thank it was a code duplication: in `gcs_set_pkt_size` GCS_PARAMS_MAX_PKT_SIZE is already set. But now I rechecked and seen that's not true: we set result of `gcs_core_set_pkt_size` to params and config, but here we set pkt_size anyway.

I'll remove this part, but I'm not sure this is correct behaviour, I'll look in.

File: gcs/src/gcs.cpp (hunk 5 above)
Comment (voodam) [reply gcs/src/gcs.cpp]: In all other param setters here we update the config. I don't know how it work in details, because `conn->config` isn't used in GCS. Either this is purposeless and we should delete updating it everywhere in gcs.cpp setters, or this config is shared between modules and hence should be fresh. But I did this for uniformity with other setters.

File: gcs/src/gcs_sm.cpp (hunk 3 above)
Comment (voodam) [reply gcs/src/gcs_sm.cpp]: Yes, agree with that

File: gcs/src/gcs_sm.hpp (hunk 7 above)
Comment (kamil-holubicki) [reply gcs/src/gcs_sm.hpp]: Please check https://stackoverflow.com/a/34034869

File: gcs/src/gcs_sm.cpp (hunk 2 above)
Comment (voodam) [reply gcs/src/gcs_sm.cpp]: Done

File: gcs/src/gcs_sm.hpp (hunk 7 above)
Comment (voodam) [reply gcs/src/gcs_sm.hpp]: Done

File: gcs/src/gcs_sm.cpp (hunk 4 above)
Comment (voodam) [reply gcs/src/gcs_sm.cpp]: Done

File: gcs/src/gcs_sm.cpp (hunk 1 above)
Comment (voodam) [reply gcs/src/gcs_sm.cpp]: Done

File: gcs/src/gcs.cpp (hunk 5 above)
Comment (voodam) [reply gcs/src/gcs.cpp]: Anyway, reverted this, because it's from upstream and don't seem significant.

File: gcs/src/gcs_sm.cpp [hunk 8]
```diff
@@ -137,9 +139,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+gcs_sm_paused_in_window_get (gcs_sm_t* const sm, long long window_ns,
+                             bool erase_outdated)
+{
+    assert(sm->fc_auto_evict);
+    long long now = gu_time_monotonic();
+    long long window_start_ns = now - window_ns;
+
+    if (gu_unlikely(gu_mutex_lock (&sm->lock))) abort();
+
+    auto & window = sm->pauses_window;
```
Comment (dutow) [inline gcs/src/gcs_sm.cpp]: Formatting

File: gcs/src/gcs_sm.cpp [hunk 9]
```diff
@@ -137,9 +139,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+    auto size = window.size();
+    assert(size % 2 == sm->pause);
+
+    long long total_ns = 0;
+    ssize_t i;
+    // paused => handle the last pause separately, start from the next pair
+    if (sm->pause) {
+        assert(window[size-1] == sm->stats.pause_start);
+        total_ns += now - sm->stats.pause_start;
```
Comment (dutow) [inline gcs/src/gcs_sm.cpp]: Why +=? It's set to be 0 just a few lines before.

File: gcs/src/gcs_sm.cpp [hunk 10]
```diff
@@ -137,9 +139,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+    sm->pauses_window.~vector();
     gu_free (sm);
 }
 
+long long
+gcs_sm_paused_in_window_get (gcs_sm_t* const sm, long long window_ns,
+                             bool erase_outdated)
+{
+    assert(sm->fc_auto_evict);
+    long long now = gu_time_monotonic();
```
Comment (dutow) [inline gcs/src/gcs_sm.cpp]: Optional: Many of these local variables could be const.

File: gcs/src/gcs_sm.cpp [hunk 11]
```diff
@@ -137,9 +139,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
...
+    assert(size % 2 == sm->pause);
+
+    long long total_ns = 0;
+    ssize_t i;
+    // paused => handle the last pause separately, start from the next pair
+    if (sm->pause) {
+        assert(window[size-1] == sm->stats.pause_start);
+        total_ns += now - sm->stats.pause_start;
+        i = size - 2;
+    // not paused => num pauses == num conts, just start from the end
```
Comment (dutow) [inline gcs/src/gcs_sm.cpp]: Strange indentation. Maybe should be at the end of next line?

File: gcs/src/gcs.cpp [hunk 12]
```diff
@@ -380,7 +394,7 @@ gcs_create (gu_config_t* const conf, gcache_t* const gcache,
         goto recv_q_failed;
     }
 
-    conn->sm = gcs_sm_create(1<<16, 1);
+    conn->sm = gcs_sm_create(1<<16, 1, conn->params.fc_auto_evict_window);
```
Comment (dutow) [inline gcs/src/gcs.cpp]: Please try to avoid implicit casts, "fc_auto_evict_window > 0" would make it clearer that this is a bool parameter.

**Review by dutow — COMMENTED** [review]
Looks good, I only added a few minor comments.

File: gcs/src/gcs.cpp (hunk 12 above)
Comment (voodam) [reply gcs/src/gcs.cpp]: Fixed all in separate commit

**Comment by kamil-holubicki** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/1602/

**Comment by kamil-holubicki** [comment]
Looks fine. Last thing: is it possible to write MTR test for this?

**Comment by voodam** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/1606/
Upd.: wrong link

**Comment by voodam** [comment]
> Looks fine. Last thing: is it possible to write MTR test for this?

Here: https://github.com/percona/percona-xtradb-cluster/pull/1783. Also fixed MTR all issue related tests, restarted them.

File: gcs/src/gcs.cpp [hunk 13]
```diff
@@ -993,6 +1037,23 @@ gcs_handle_flow_control (gcs_conn_t*                conn,
...
     }
 
+    if (conn->params.fc_auto_evict_window > 0) {
+        const long long paused =
+            gcs_sm_paused_in_window_get(conn->sm, conn->fc_auto_evict_window_ns,
+                                        !(conn->local_act_id % 10)); // erase not very often
+        gu_debug ("FC auto evict: paused for %lld ns.", paused);
+
+        if (paused >= conn->fc_auto_evict_threshold_ns) {
+            gu_info ("FC auto evict: pause limit exceeded (%f from %f sec. in %f sec. window). "
```
Comment (venkatesh-prasad-v) [inline gcs/src/gcs.cpp]: Can we rephrase this message to something readable like
```
 Triggering automatic eviction of the node since flow control pause limit exceeded from %f to %f in %f sec window. The node will be aborted.
```

Comment (voodam) [reply gcs/src/gcs.cpp]: Done

**Review by venkatesh-prasad-v — APPROVED** [review]

**Review by kamil-holubicki — APPROVED** [review]
